src/Basic/Loader.hpp

```diff
@@ -0,0 +1,255 @@
+                         get_opt<float>(e, "radius"),
+                         get_opt<glm::vec2>(e, "curveFrom"),
+                         get_opt<glm::vec2>(e, "curveTo"),
+                         get_opt<int>(e, "cornerStyle"));
+  }
+  return contour;
+}
+
+inline std::shared_ptr<ImageNode> fromImage(const nlohmann::json& j)
+{
+  auto p = std::make_shared<ImageNode>(j["name"]);
+  fromObjectCommonProperty(j, p.get());
+  p->setImage(j["imageFileName"]);
+  if (auto it = j.find("fillReplacesImage"); it != j.end())
+  {
+    p->setReplacesImage(it.value());
+  }
+  return p;
+}
+
+inline std::shared_ptr<SymbolInstanceNode> fromSymbolInstance(const nlohmann::json& j)
+{
+  auto ins = std::make_shared<SymbolInstanceNode>(j["name"]);
+  ins->symbolID = j["symbolID"];
+  fromObjectCommonProperty(j, ins.get());
+  return ins;
+}
+
+inline std::shared_ptr<TextNode> fromText(const nlohmann::json& j)
+{
+  return std::make_shared<TextNode>("Text");
+}
+
+inline std::shared_ptr<PathNode> fromPath(const nlohmann::json& j)
+{
+  auto p = std::make_shared<PathNode>(j["name"]);
+  fromObjectCommonProperty(j, p.get());
+  const auto shape = j["shape"];
+  p->shape.windingRule = shape["windingRule"];
+  for (const auto& subshape : shape["subshapes"])
+  {
+    const auto blop = subshape["booleanOperation"];
+    p->shape.subshape.blop = blop;
+    const auto geo = subshape["subGeometry"];
+    const auto klass = geo["class"];
+    if (klass == "contour")
+    {
+      p->shape.subshape.contour = fromContour(geo);
+    }
+    else if (klass == "path")
+    {
+      p->pushChildBack(fromPath(geo));
+    }
+    else if (klass == "image")
+    {
+      p->pushChildBack(fromImage(geo));
+    }
+    else if (klass == "text")
+    {
+      p->pushChildBack(fromText(geo));
+    }
+    else if (klass == "group")
+    {
+      p->pushChildBack(fromGroup(j));
+    }
+    else if (klass == "symbolInstance")
+    {
+      p->pushChildBack(fromSymbolInstance(geo));
+    }
+  }
+  return p;
+}
+
+inline std::shared_ptr<PaintNode> fromObject(const nlohmann::json& j)
+{
+  std::shared_ptr<PaintNode> ro;
+  auto klass = j["class"];
+  if (klass == "group")
+  {
+    ro = fromGroup(j);
+  }
+  else if (klass == "path")
+  {
+    ro = fromPath(j);
+  }
+  else if (klass == "image")
+  {
+    ro = fromImage(j);
+  }
+  else if (klass == "text")
+  {
+    ro = fromText(j);
+  }
+  else if (klass == "symbolInstance")
+  {
+    ro = fromSymbolInstance(j);
+  }
+  fromObjectCommonProperty(j, ro.get());
+  return ro;
+}
+
+inline std::shared_ptr<GroupNode> fromGroup(const nlohmann::json& j)
+{
+  auto p = std::make_shared<GroupNode>("group");
+  // init group properties
+  fromObjectCommonProperty(j, p.get());
+  for (const auto& c : j["childObjects"])
+  {
+    p->pushChildBack(fromObject(c));
+  }
+  return p;
+}
+
+inline std::shared_ptr<PaintNode> fromLayer(const nlohmann::json& j)
+{
+  auto p = std::make_shared<PaintNode>("Layer", VGG_LAYER);
```

Comment:
Why is this special with enum?

Reply:
It's not necessary to add a ```LayerNode```  because there is no sophisticated functions about the layer. Just use a base class with a flag is enough. If more function need to deal with layer, a concrete class will replace it.